openedx/core/djangoapps/catalog/utils.py

```diff
@@ -239,6 +242,47 @@ def get_course_runs_for_course(course_uuid):
     else:
         return []
 
+def get_visible_course_runs_for_course(entitlement, course_uuid):
+    """
+    We only want to show courses for a particular entitlement that:
+
+    1) Are currently running or in the future
+    2) A user can enroll in
+    3) A user can upgrade in
+    4) Are published
+    """
+    course_runs_for_course = get_course_runs_for_course(course_uuid)
+    enrollable_course_runs = []
+
+    # Only show published course runs that can still be enrolled and upgraded
+    now = datetime.datetime.now(UTC)
+    for course_run in course_runs_for_course:
+        # Only courses that have not ended will be displayed
+        run_start = course_run.get('start')
+        run_end = course_run.get('end')
+        is_running = run_start and run_end and datetime_parse(run_end) > now
+
+        # Only courses that can currently be enrolled in will be displayed
+        enrollment_start = course_run.get('enrollment_start')
+        enrollment_end = course_run.get('enrollment_end')
+        can_enroll = ((not enrollment_start or datetime_parse(enrollment_start) < now)
+                      and (not enrollment_end or (datetime_parse(enrollment_end) > now)))
+
+        # Only upgrade-able courses will be displayed
+        upgrade_deadline = next(iter(
+            [seat.get('upgrade_deadline') for seat in course_run.get('seats')
+             if seat.get('type') == entitlement.mode]), None
+        )
```

Comment:
List comprehensions should only be used when building and returning a list.  This seems like overkill, and will cause code readability issues in the future for what is simply an attempt to get a date. Also it will iterate over all the seats when you only need to get the first matching seat_type for the mode.

If I am misunderstanding the goal with this code please let me know.

Consider the following sudo code simplification
```
        # Only upgrade-able courses will be displayed
        for seat in course_run.get('seats', []):
            if seat.get('type') == entitlement.mode:
                upgrade_deadline = seat.get('upgrade_deadline', None)
                break
```

Reply:
In short I think this is one case where a simple for loop with a break is better than a list comprehension.